feat: add audit log search and get endpoints

danielkelemen opened the pull request:
## Description

Add audit log search and get endpoints

## Related issues

closes https://github.com/camunda/camunda/issues/41133

danielkelemen committed:
feat: add audit log search and get endpoints

closes https://github.com/camunda/camunda/issues/41133

nicpuppa commented on zeebe/gateway-protocol/src/main/proto/v2/audit-logs.yaml:
```diff
@@ -0,0 +1,329 @@
+          $ref: 'common-responses.yaml#/components/responses/Forbidden'
+        "500":
+          description: An internal error occurred while processing the request.
+      x-eventually-consistent: true
```
🔧 Please move this at the top, before tags

danielkelemen committed:
refactor: review feedback

danielkelemen commented on zeebe/gateway-protocol/src/main/proto/v2/audit-logs.yaml:
```diff
@@ -0,0 +1,329 @@
+          allOf:
+            - $ref: '#/components/schemas/AuditLogEntityTypeEnum'
+          description: The type of the entity.
+        operationType:
```
This could be a long enum, it contains all the possible values of Intents.

@koevskinikola, what do you think?

nicpuppa reviewed (approved):
Just a small suggestion please look at it. 

Pre-approving from a specification perspective, wating for @koevskinikola from the implementation point of view 👍

danielkelemen committed:
feat: improve audit log schema enums

tmetzke commented on zeebe/gateway-protocol/src/main/proto/v2/audit-logs.yaml:
```diff
@@ -0,0 +1,312 @@
+            - $ref: 'keys.yaml#/components/schemas/AuditLogOperationKey'
+          description: The unique operation key of the audit log entry.
+        entityKey:
+          type: string
```
💭 Could this be a reference to the `LongKey`, actually? It will always point to keys that adhere to that schema, right?

danielkelemen replied:
The plan is that for identity related operations this will contain the email address for instance.

[From the solution design](https://docs.google.com/document/d/1COzShIcqFqZlq6ZRfuL9VeHD7oNvVd8vhnezxGlYTU0/edit?tab=t.vpvtku4f2lqx#heading=h.kd9atrl0nn0s): `May be a Long key, or an Identity-related String ID (email)`

I'd keep it `string` for now.

tmetzke replied:
Well, we can clearly state that it only makes sense to always consider both, I guess.

danielkelemen replied:
Yep, let's go with this for now. If we discover something during identity milestone, we can extend it.

tmetzke commented on zeebe/gateway-protocol/src/main/proto/v2/audit-logs.yaml:
```diff
@@ -0,0 +1,312 @@
+          allOf:
+            - $ref: '#/components/schemas/AuditLogBatchOperationTypeEnum'
+          description: The type of batch operation performed, if this is part of a batch.
+        timestamp:
```
❌ We try to avoid raw timestamps in the API if possible and rather use date-time string. Can we do this here as well?

tmetzke commented on zeebe/gateway-protocol/src/main/proto/v2/keys.yaml:
```diff
@@ -149,6 +149,15 @@ components:
       format: int64
       minimum: 1
 
+    AuditLogOperationKey:
```
Nice, thanks for clarifying, makes sense then to keep it here 👍

danielkelemen commented on zeebe/gateway-protocol/src/main/proto/v2/audit-logs.yaml:
```diff
@@ -0,0 +1,312 @@
+        timestamp:
+          description: The timestamp search filter.
+          allOf:
+            - $ref: 'filters.yaml#/components/schemas/DateTimeFilterProperty'
```
Yes, it should be a date-time.

tmetzke commented on zeebe/gateway-protocol/src/main/proto/v2/audit-logs.yaml:
```diff
@@ -0,0 +1,312 @@
+      description: Audit log item.
+      type: object
+      properties:
+        operationKey:
```
I'm afraid users will easily mix audit logs up with batch operations if we call them "operations", to be honest. It's fine to keep the batch operation attributes as they are, I would say.

I could imagine the following taxonomy:
* It's an audit log (instance).
* It has an `auditLogKey` as a primary key. This is in our [REST guidelines](https://docs.camunda.io/docs/next/apis-tools/orchestration-cluster-api-rest/orchestration-cluster-api-rest-overview/#naming-conventions).
* It has a type, so maybe an `auditLogType` to clearly differentiate it in this case.
* It references other entities and their properties, like `batchOperationKey`, `batchOperationType`, and so on. They clearly state with entity they belong to.

danielkelemen committed:
refactor: improve audit log api

- rename key to auditLogKey
- reuse BatchOperationTypeEnum
- fix timestamp type
- add AuditLogKeyFilterProperty

danielkelemen committed:
feat: add advanced search filters for audit log properties

tmetzke reviewed (approved):
👍 Looks good to me overall.
🔧 Some minor cleanups left to follow up on.

danielkelemen committed:
feat: expand sortable fields in audit log schema

danielkelemen committed:
feat: add advanced category filter to audit log schema

koevskinikola commented on zeebe/gateway-protocol/src/main/proto/v2/audit-logs.yaml:
```diff
@@ -0,0 +1,505 @@
+        - Audit Log
+      operationId: getAuditLog
+      summary: Get audit log
+      description: Get an audit log by auditLogKey.
```
🔧 We get a single entry
```suggestion
      description: Get an audit log entry by auditLogKey.
```

koevskinikola commented on zeebe/gateway-protocol/src/main/proto/v2/audit-logs.yaml:
```diff
@@ -0,0 +1,505 @@
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/AuditLogResult'
```
❓ @danielkelemen we have a requirement that, if audit log tracking is disabled, we return a message that describes how to configure it, instead of an error response.

Should we model it now, or do we schedule it  when we add the configuration properties?

koevskinikola commented on zeebe/gateway-protocol/src/main/proto/v2/audit-logs.yaml:
```diff
@@ -0,0 +1,505 @@
+          allOf:
+            - $ref: 'identifiers.yaml#/components/schemas/TenantId'
+          description: The tenant ID of the audit log.
+        operationStatus:
```
❌ Let's change this to `result`, as it's only `SUCCESS` or `FAIL`, there's no other status.

koevskinikola reviewed (changes requested):
@danielkelemen, looks good, I only added feedback to some properties that were renamed when the solution architecture was approved.

I also think adding a separate `entityId` field will be useful for Identity properties. Otherwise, it will be difficult for us to distinguish between Spring and Long keys/IDs.

tmetzke commented on zeebe/gateway-protocol/src/main/proto/v2/audit-logs.yaml:
```diff
@@ -0,0 +1,505 @@
+        - USER_TASK
+        - ADMIN
+
+    ## Filters
```
Do you mean the property types they are referencing? Because they already exist in the files they reference.

danielkelemen committed:
refactor: improve audit log api

- rename status to result
- drop result and actorType advanced filters
- few smaller improvements

koevskinikola reviewed (approved):
LGTM @danielkelemen 🚀
